python/salpytools/salpylib.py

```diff
@@ -141,37 +141,65 @@ def validate_transition(self, new_state):
 
     def execute_command(self, command):
         """This method delegates commands recieved by a DDSController to the
-        state. 
+        state.
 
         The model is passed so that the State object may call methods on
         it. Also the state is stored on the model only as a string
         representation. The actual state object is stored on this context
-        object as self.states. 
+        object as self.states.
 
         Attributes:
             command: A string representation of the command recieved by a
             DDSController object.
         """
 
-        current_state = self.states[self.model.state]
-
         if command == "ENTERCONTROL":
-            current_state.enter_control()
+
+            # Get the current state
+            current_state = self.states[self.model.state]
+            # Call exit methods, killing threads perhaps
+            current_state.exit(self.model)
+            # Call our "change state" method
+            current_state.enter_control(self.model)
```

Comment:
This should probably happen after you current_state is set to the new state. Same as for all other commands.